sdk/core/core-tracing/review/core-tracing.api.md

```diff
@@ -5,183 +5,116 @@
-
-// @public
-export function getTracer(): Tracer;
-
-// @public
-export function getTracer(name: string, version?: string): Tracer;
-
-// @public
-export type HrTime = [number, number];
-
-// @public
-export function isSpanContextValid(context: SpanContext): boolean;
+export type SpanStatus = {
+    status: "success";
+} | {
+    status: "error";
+    error: Error | string;
+};
 
 // @public
-export interface Link {
-    attributes?: SpanAttributes;
-    context: SpanContext;
+export interface TracingClient {
+    createRequestHeaders(spanId: TracingSpanIdentifier): Record<string, string>;
+    parseTraceparentHeader(traceparentHeader: string): TracingSpanIdentifier | undefined;
+    startSpan<Options extends {
+        tracingOptions?: OperationTracingOptions;
+    }>(name: string, operationOptions?: Options, spanOptions?: TracingSpanOptions): {
+        span: TracingSpan;
+        tracingContext: TracingContext;
+        updatedOptions: Options;
+    };
+    withContext<CallbackArgs extends unknown[], Callback extends (...args: CallbackArgs) => ReturnType<Callback>>(context: TracingContext, callback: Callback, callbackThis?: ThisParameterType<Callback>, ...callbackArgs: CallbackArgs): ReturnType<Callback>;
+    withSpan<Options extends {
+        tracingOptions?: OperationTracingOptions;
+    }, Callback extends (updatedOptions: Options, span: Omit<TracingSpan, "end">) => ReturnType<Callback>>(name: string, callback: Callback, operationOptions?: Options, spanOptions?: TracingSpanOptions, callbackThis?: ThisParameterType<Callback>): Promise<ReturnType<Callback>>;
 }
 
 // @public
-export interface OperationTracingOptions {
-    tracingContext?: Context;
+export interface TracingClientOptions {
+    namespace: string;
+    packageInformation: {
+        name: string;
+        version?: string;
+    };
 }
 
 // @public
-export function setSpan(context: Context, span: Span): Context;
-
-// @public
-export function setSpanContext(context: Context, spanContext: SpanContext): Context;
-
-// @public
-export interface Span {
-    addEvent(name: string, attributesOrStartTime?: SpanAttributes | TimeInput, startTime?: TimeInput): this;
-    end(endTime?: TimeInput): void;
-    isRecording(): boolean;
-    recordException(exception: Exception, time?: TimeInput): void;
-    setAttribute(key: string, value: SpanAttributeValue): this;
-    setAttributes(attributes: SpanAttributes): this;
-    setStatus(status: SpanStatus): this;
-    spanContext(): SpanContext;
-    updateName(name: string): this;
+export interface TracingContext {
+    deleteValue(key: symbol): TracingContext;
+    getValue(key: symbol): unknown;
+    setValue(key: symbol, value: unknown): TracingContext;
 }
 
 // @public
-export interface SpanAttributes {
-    [attributeKey: string]: SpanAttributeValue | undefined;
+export interface TracingSpan {
+    end(): void;
+    isRecording(): boolean;
+    setAttribute(name: string, value: unknown): void;
+    setStatus(status: SpanStatus): void;
```

Comment:
we'd also want a way to set exception. We can do it by adding a new parameter here or a new method like `set(add/record)Exception`

Reply:
So what I was thinking here is that the user will make a single call to setStatus and we will then call the right APIs under the hood.

Here's some examples:

- Success
  - User calls `setStatus({ status: "success" })`
  - OTelInstrumenter calls `setStatus({ code: SpanStatusCode.OK })`
- Error
  - User calls `setStatus({ status: "error", error: err })` // err is the exception that was caught, and is optional
  - OTelInstrumenter calls: `setStatus({ code: SpanStatusCode.ERROR, message: err.message }); recordException(err)`

This way the caller never has to remember to call `recordException` and we will handle it as an implementation detail - what do you think?

Reply:
Sorry just to back up as I re-read the comment - SpanStatus is a union type that can be either success or error - in the case of an error status the user can also pass in the exception itself or a message and we will then call recordException.

Reply:
There are cases where you want to record multiple exceptions without changing the span status - so we do need a separate way to recordException...